RFC-0030-Consolidate-Elastic-and-TorchX.md

```diff
@@ -0,0 +1,262 @@
+
+The following table summarizes when each tool was introduced to PyTorch and how they relate to each other.
+
+| **PyTorch Version**  | `torch.distributed.launch`                                   | `torch.distributed.run` | `torchrun`                                                                                                      |
+|----------------------|--------------------------------------------------------------|-------------------------|-----------------------------------------------------------------------------------------------------------------|
+| <1.9                 | Uses `torch.multiprocessing`                                 | --                      | --                                                                                                              |
+| 1.9                  | Wraps `torch.distributed.run` with backwards compatible args | Uses `torchelastic`     | --                                                                                                              |
+| 1.10                 | [same as above]                                              | [same as above]         | CLI [console_script](https://github.com/pytorch/pytorch/blob/master/setup.py#L980) to `torch.distributed.run` |
+
+
+### TorchX
+
+[TorchX](https://pytorch.org/torchx)
+on the other hand is a **job launcher** that can submit PyTorch scripts as a job onto various remote schedulers
+such as SLURM, Kubernetes, AWS/GCP/Azure Batch, etc. Similar to TorchElastic prior to `torch-1.9`,
+TorchX is currently in the PyTorch GitHub organization but under the [pytorch/torchx](https://github.com/pytorch/torchx)
+repository.
+
+
+## **Motivation**
+
+1. **Uniform user experience (UX)** 
+     1. Both `torchrun` and `torchx` **launch** PyTorch scripts 
+     2. For instance, `torchx run -s local dist.ddp -j 1x2 train.py` is basically equivalent to `torchrun --nnodes 1 --nproc_per_node 2 train.py`
+2. **More native built-in support in PyTorch for launching large-scale, long-running, fault-tolerant distributed jobs**
+     1. In practice running large-scale jobs reliably requires coordination between the scheduler/infrastructure (controlled by torchx)
+         and the process manager (torchelastic)
+     2. Example: dealing with host failures in a job:
+        1. torchelastic alone cannot deal with host failures. The job needs to be configured (done via torchx) for
+           the scheduler to deal with host failures (e.g. replace the failed host).
+        2. Conversely, the scheduler replacing failed hosts for a job is not enough, the job itself needs to be authored
+           in such a way that surviving hosts can wait-for a failed host to be replaced then admit the newly replace host
+           into the worker group (provided by torchelastic).
+3. **Out-of-the-box support for job launchers in PyTorch** 
+    1. Without additional installs, PyTorch users can run PyTorch scripts both locally and remotely onto widely used schedulers 
```

Comment:
Actually if this is all living in core it might make sense to have all schedulers out core with maybe some exception for the very well maintained or popular ones like K8, that way you can standardize a job launching interface without having to contribute more experimental schedulers in core where they may likely rot

Reply:
yea I think it makes sense to include the schedulers that are not CSP specific (which would be SLURM and Kubernetes - SLURM doesn't even have a build-time dep since it doesn't have a python client), and keep the CSP ones (e.g. AWS Batch) in torchx so that eventually we can make the relevant service teams own the integration.